src/main/java/com/shapesecurity/shift/validator/ValidationContext.java

```diff
@@ -67,82 +71,229 @@ private ValidationContext(@NotNull ConcatList<ValidationError> freeBreakStatemen
 
   public ValidationContext addFreeBreakStatement(@NotNull ValidationError statement) {
     return new ValidationContext(
-        this.freeBreakStatements.append(ConcatList.single(statement)),
-        this.freeContinueStatements, this.usedLabelNames, this.freeJumpTargets, this.freeReturnStatements, this.errors,
-        this.strictErrors);
+        this.freeBreakStatements.append1(statement),
+        this.freeContinueStatements,
+        this.usedLabelNames,
+        this.freeJumpTargets,
+        this.freeReturnStatements,
+        this.errors,
+        this.strictErrors
+    );
   }
 
   public ValidationContext clearFreeBreakStatements() {
     return new ValidationContext(
         ConcatList.empty(), //empty freeBreakStatements
-        this.freeContinueStatements, this.usedLabelNames, this.freeJumpTargets, this.freeReturnStatements, this.errors,
-        this.strictErrors);
+        this.freeContinueStatements,
+        this.usedLabelNames,
+        this.freeJumpTargets,
+        this.freeReturnStatements,
+        this.errors,
+        this.strictErrors
+    );
   }
 
-  public ValidationContext addFreeContinueStatement(@NotNull ValidationError statement) {
+  public ValidationContext addFreeContinueStatement(@NotNull ContinueStatement statement) {
     return new ValidationContext(
-        this.freeBreakStatements, this.freeContinueStatements.append(
-        ConcatList.single(
-            statement)), this.usedLabelNames, this.freeJumpTargets, this.freeReturnStatements, this.errors,
-        this.strictErrors);
+        this.freeBreakStatements,
+        this.freeContinueStatements.append1(new ValidationError(statement,
+            "Continue statement must be inside a recursive loop")),
```

Comment:
Can you change this "recursive loop" message? It must be in an iteration statement.